### PR TITLE
[GrandCentral] FieldID support using probe

### PR DESCRIPTION
**Pull request by uenoku**
This commit adds `fieldID` member to InnerRefAttr something like,
`#hw.innerNameRef<@DUT::@c, 4>`.
It allows users to specify specific elements of an aggregate 
value in verbatim operations.

Implementation wise, `EmitterBase::emitTextWithSubstutions` is modified
to emit the accociated path if the fieldID is not zero.

Will close https://github.com/llvm/circt/issues/2409

### CODE REVIEW

File: include/circt/Dialect/HW/HWAttributesNaming.td [hunk 1]
```diff
@@ -16,16 +16,28 @@ def InnerRefAttr : AttrDef<HWDialect, "InnerRef"> {
     This works like a symbol reference, but to a name inside a module.
   }];
   let mnemonic = "innerNameRef";
-  let parameters = (ins "::mlir::StringAttr":$module, "::mlir::StringAttr":$name);
+  let parameters = (ins "::mlir::StringAttr":$module, "::mlir::StringAttr":$name,
+                        "unsigned":$fieldID);
```
Comment (uenoku) [inline include/circt/Dialect/HW/HWAttributesNaming.td:20]: I wonder it is possible to make "fieldID" to optional parameter whose default value is zero.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 2]
```diff
@@ -492,6 +492,52 @@ static bool isOkToBitSelectFrom(Value v) {
   return false;
 }
 
+/// Construct a access path which points at the element given by `fieldID`.
+/// TODO: This function should be removed once HW/SV dialects implement
+/// fieldID utilities like FIRRTL dialect.
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:497]: Can you open an issue to track this? I agree that this should be short lived while support for fieldID is added to HW/SV.

**Review by seldridge — APPROVED** [review]

File: include/circt/Dialect/HW/HWAttributesNaming.td [hunk 3]
```diff
@@ -16,16 +16,28 @@ def InnerRefAttr : AttrDef<HWDialect, "InnerRef"> {
...
   let builders = [
-    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$module, "::mlir::StringAttr":$name),[{
-      return get(module.getContext(), module, name);
+    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$module,
+                                        "::mlir::StringAttr":$name,
+                                        "unsigned":$fieldID),[{
```
Comment (seldridge) [inline include/circt/Dialect/HW/HWAttributesNaming.td:24]: This should probably be an `I32Attr` or `APIntAttr` instead of an `unsigned`.

Comment (uenoku) [reply include/circt/Dialect/HW/HWAttributesNaming.td:24]: It seems `I32Attr` or `APIntATttr` cannot be used as a parameter of attributes. I think it is possible to use mlir::IntegerAttr but  I feel there is no reason to use IntegerAttr here.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp:497]: Opened https://github.com/llvm/circt/issues/2419

**Comment by darthscsi** [comment]
I'm not sure I like the idea of innerNameRef having a fieldID.  That seems the problem of the user.  It seems innerNameRefs should only refer to namable things and a field of a variable is not namable.

**Comment by uenoku** [comment]
> I'm not sure I like the idea of innerNameRef having a fieldID. That seems the problem of the user. It seems innerNameRefs should only refer to namable things and a field of a variable is not namable.

That makes sense. Actually it is reasonable to separate the attribute to refer the path like, `InnerFieldRef(innerNameRef, fieldID)` or something.

**Comment by uenoku** [comment]
Separated into `innerFieldRefAttr`.  This might be arguable that innerFieldRef only prints a path, for example `#hw.innerFieldRef<@Foo::@wire, 2>` would be printed as `.[1]`.

**Comment by uenoku** [comment]
I would like to merge this if there is no problem.

File: test/Dialect/FIRRTL/grand-central.mlir [hunk 4]
```diff
@@ -85,11 +85,12 @@ firrtl.circuit "InterfaceGroundType" attributes {
 // CHECK-SAME:   @InterfaceGroundType
 // CHECK-SAME:   #hw.innerNameRef<@InterfaceGroundType::@dut>
 // CHECK-SAME:   #hw.innerNameRef<@DUT::@b>
-// CHECK-NEXT: sv.verbatim "assign {{[{][{]0[}][}]}}.baz = {{[{][{]1[}][}]}}.{{[{][{]2[}][}]}}.{{[{][{]3[}][}]}}[1].d;"
+// CHECK-NEXT: sv.verbatim "assign {{[{][{]0[}][}]}}.baz = {{[{][{]1[}][}]}}.{{[{][{]2[}][}]}}.{{[{][{]3[}][}]}}{{[{][{]4[}][}]}};"
```
Comment (seldridge) [inline test/Dialect/FIRRTL/grand-central.mlir:88]: This continues to look great. 💯

**Review by seldridge — APPROVED** [review]
lgtm

File: test/Dialect/FIRRTL/grand-central.mlir (hunk 4 above)
Comment (darthscsi) [inline test/Dialect/FIRRTL/grand-central.mlir:88]: In the module, an XMR followed by a index op with the result passed to the verbatim would be equivalent and not require new operations.

**Review by darthscsi — CHANGES_REQUESTED** [review]
Please see if you can achieve what you need with existing ops, like XMRs and subfieldaccess.

**Comment by uenoku** [comment]
> Please see if you can achieve what you need with existing ops, like XMRs and subfieldaccess.

I didn't know about XMR. Thanks!

File: test/Dialect/FIRRTL/grand-central.mlir (hunk 4 above)
Comment (darthscsi) [reply test/Dialect/FIRRTL/grand-central.mlir:88]: Or perhaps more appropriately an subindex/subaccess in the referenced module feeding a hw.probe.

**Comment by darthscsi** [comment]
XMR is definitely the wrong thing here.  probe would be much better.

**Comment by uenoku** [comment]
Using probe makes sense. Then I'll wait for probe lowering support.

**Review by seldridge — APPROVED** [review]
Great suggestion to use `firrtl.probe` here. This is way cleaner.

**Comment by uenoku** [comment]
Close this for now since this looks quite old.

**Pull request closed without merging**